### PR TITLE
New Resource:`azurerm_site_recovery_hyperv_replication_policy`

### PR DESCRIPTION
**Pull request by ziyeqf**
test
---
```
terraform-provider-azurerm ❯ TF_ACC=1 go test -v ./internal/services/recoveryservices -run=TestAccSiteRecoveryReplicationPolicyHyperV_ -timeout=600m
=== RUN   TestAccSiteRecoveryReplicationPolicyHyperV_basic
=== PAUSE TestAccSiteRecoveryReplicationPolicyHyperV_basic
=== RUN   TestAccSiteRecoveryReplicationPolicyHyperV_noSnapshots
=== PAUSE TestAccSiteRecoveryReplicationPolicyHyperV_noSnapshots
=== CONT  TestAccSiteRecoveryReplicationPolicyHyperV_basic
=== CONT  TestAccSiteRecoveryReplicationPolicyHyperV_noSnapshots
--- PASS: TestAccSiteRecoveryReplicationPolicyHyperV_noSnapshots (481.83s)
--- PASS: TestAccSiteRecoveryReplicationPolicyHyperV_basic (492.59s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/recoveryservices      493.819s
```

### CODE REVIEW

File: internal/services/recoveryservices/site_recovery_replication_policy_hyperv_resource.go [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+	"github.com/hashicorp/terraform-provider-azurerm/internal/sdk"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/services/recoveryservices/validate"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/tf/pluginsdk"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/tf/validation"
+)
+
+type ReplicationPolicyHyperVModel struct {
+	Name                                          string `tfschema:"name"`
+	ResourceGroupName                             string `tfschema:"resource_group_name"`
+	RecoveryVaultName                             string `tfschema:"recovery_vault_name"`
```
Comment (stephybun) [inline internal/services/recoveryservices/site_recovery_replication_policy_hyperv_resource.go:21]: We can replace these with one property: 
```suggestion
	RecoveryVaultId                             string `tfschema:"recovery_vault_id"`
```

File: internal/services/recoveryservices/site_recovery_replication_policy_hyperv_resource.go [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+			Type:         pluginsdk.TypeInt,
+			Required:     true,
+			ForceNew:     false,
+			ValidateFunc: validation.IntBetween(0, 12),
+		},
+		"replication_interval_in_seconds": {
+			Type:         pluginsdk.TypeInt,
+			Required:     true,
+			ForceNew:     false,
+			ValidateFunc: validation.IntInSlice([]int{30, 300}),
```
Comment (stephybun) [inline internal/services/recoveryservices/site_recovery_replication_policy_hyperv_resource.go:63]: `ForceNew: false` can be removed
```suggestion
		"recovery_point_retention_in_hours": {
			Type:         pluginsdk.TypeInt,
			Required:     true,
			ValidateFunc: validation.IntBetween(0, 24),
		},
		"application_consistent_snapshot_frequency_in_hours": {
			Type:         pluginsdk.TypeInt,
			Required:     true,
			ValidateFunc: validation.IntBetween(0, 12),
		},
		"replication_interval_in_seconds": {
			Type:         pluginsdk.TypeInt,
			Required:     true,
			ValidateFunc: validation.IntInSlice([]int{30, 300}),
```

File: internal/services/recoveryservices/site_recovery_replication_policy_hyperv_resource.go [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+
+			existing, err := client.Get(ctx, id)
+			if err != nil {
+				// NOTE: Bad Request due to https://github.com/Azure/azure-rest-api-specs/issues/12759
+				if !response.WasNotFound(existing.HttpResponse) && !wasBadRequestWithNotExist(existing.HttpResponse, err) {
+					return fmt.Errorf("checking presence %s: %+v", plan.Name, err)
+				}
+			}
+
+			if existing.Model != nil && existing.Model.Id != nil && *existing.Model.Id != "" {
```
Comment (stephybun) [inline internal/services/recoveryservices/site_recovery_replication_policy_hyperv_resource.go:102]: It's sufficient to just check whether the Model exists in this case
```suggestion
			if existing.Model != nil {
```

File: internal/services/recoveryservices/site_recovery_replication_policy_hyperv_resource.go [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+			subscriptionId := metadata.Client.Account.SubscriptionId
+			client := metadata.Client.RecoveryServices.ReplicationPoliciesClient
+
+			id := replicationpolicies.NewReplicationPolicyID(subscriptionId, plan.ResourceGroupName, plan.RecoveryVaultName, plan.Name)
+
+			existing, err := client.Get(ctx, id)
+			if err != nil {
+				// NOTE: Bad Request due to https://github.com/Azure/azure-rest-api-specs/issues/12759
+				if !response.WasNotFound(existing.HttpResponse) && !wasBadRequestWithNotExist(existing.HttpResponse, err) {
+					return fmt.Errorf("checking presence %s: %+v", plan.Name, err)
```
Comment (stephybun) [inline internal/services/recoveryservices/site_recovery_replication_policy_hyperv_resource.go:98]:
```suggestion
					return fmt.Errorf("checking for existing %s: %+v", plan.Name, err)
```

File: internal/services/recoveryservices/site_recovery_replication_policy_hyperv_resource.go [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+					ProviderSpecificInput: &replicationpolicies.HyperVReplicaAzurePolicyInput{
+						RecoveryPointHistoryDuration:                  &plan.RecoveryPointRetentionInHours,
+						ApplicationConsistentSnapshotFrequencyInHours: &plan.ApplicationConsistentSnapshotFrequencyInHours,
+						ReplicationInterval:                           &plan.CopyFrequency,
+					},
+				},
+			}
+			err = client.CreateThenPoll(ctx, id, parameters)
+			if err != nil {
+				return fmt.Errorf("creating site recovery replication policy %s: %+v", id, err)
```
Comment (stephybun) [inline internal/services/recoveryservices/site_recovery_replication_policy_hyperv_resource.go:117]:
```suggestion
				return fmt.Errorf("creating %s: %+v", id, err)
```

File: internal/services/recoveryservices/site_recovery_replication_policy_hyperv_resource.go [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+	}
+}
+
+func (r ReplicationPolicyHyperVResource) Read() sdk.ResourceFunc {
+	return sdk.ResourceFunc{
+		Timeout: 5 * time.Minute,
+		Func: func(ctx context.Context, metadata sdk.ResourceMetaData) error {
+			id, err := replicationpolicies.ParseReplicationPolicyID(metadata.ResourceData.Id())
+			if err != nil {
+				return fmt.Errorf("parsing %s: %+v", metadata.ResourceData.Id(), err)
```
Comment (stephybun) [inline internal/services/recoveryservices/site_recovery_replication_policy_hyperv_resource.go:133]:
```suggestion
				return err
```

File: internal/services/recoveryservices/site_recovery_replication_policy_hyperv_resource.go [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+			}
+
+			client := metadata.Client.RecoveryServices.ReplicationPoliciesClient
+
+			resp, err := client.Get(ctx, *id)
+			if err != nil {
+				if response.WasNotFound(resp.HttpResponse) {
+					return metadata.MarkAsGone(id)
+				}
+				return fmt.Errorf("reading %s : %+v", id.String(), err)
```
Comment (stephybun) [inline internal/services/recoveryservices/site_recovery_replication_policy_hyperv_resource.go:143]:
```suggestion
				return fmt.Errorf("retrieving %s : %+v", *id, err)
```

File: internal/services/recoveryservices/site_recovery_replication_policy_hyperv_resource.go [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+		Timeout: 30 * time.Minute,
+		Func: func(ctx context.Context, metadata sdk.ResourceMetaData) error {
+			var model ReplicationPolicyHyperVModel
+			if err := metadata.Decode(&model); err != nil {
+				return fmt.Errorf("decoding %+v", err)
+			}
+
+			id, err := replicationpolicies.ParseReplicationPolicyID(metadata.ResourceData.Id())
+			if err != nil {
+				return fmt.Errorf("parsing %s: %+v", metadata.ResourceData.Id(), err)
```
Comment (stephybun) [inline internal/services/recoveryservices/site_recovery_replication_policy_hyperv_resource.go:182]:
```suggestion
				return err
```

File: internal/services/recoveryservices/site_recovery_replication_policy_hyperv_resource.go [hunk 9]
```diff
@@ -0,0 +1,244 @@
...
+			id, err := replicationpolicies.ParseReplicationPolicyID(metadata.ResourceData.Id())
+			if err != nil {
+				return err
+			}
+
+			client := metadata.Client.RecoveryServices.ReplicationPoliciesClient
+
+			err = client.DeleteThenPoll(ctx, *id)
+			if err != nil {
+				return fmt.Errorf("deleting  %s : %+v", id, err)
```
Comment (stephybun) [inline internal/services/recoveryservices/site_recovery_replication_policy_hyperv_resource.go:220]:
```suggestion
				return fmt.Errorf("deleting %s: %+v", id, err)
```

File: internal/services/recoveryservices/site_recovery_replication_policy_hyperv_resource.go [hunk 10]
```diff
@@ -0,0 +1,244 @@
...
+			return nil
+		},
+	}
+}
+
+func (r ReplicationPolicyHyperVResource) IDValidationFunc() pluginsdk.SchemaValidateFunc {
+	return replicationpolicies.ValidateReplicationPolicyID
+}
+
+func expandH2APolicyDetail(input *replicationpolicies.Policy) (out *replicationpolicies.HyperVReplicaAzurePolicyDetails, isA2A bool) {
```
Comment (stephybun) [inline internal/services/recoveryservices/site_recovery_replication_policy_hyperv_resource.go:232]: What is H2A and A2A?

File: internal/services/recoveryservices/site_recovery_replication_policy_hyperv_resource_test.go [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+
+func (t SiteRecoveryReplicationPolicyHyperVResource) Exists(ctx context.Context, clients *clients.Client, state *pluginsdk.InstanceState) (*bool, error) {
+	id, err := replicationpolicies.ParseReplicationPolicyID(state.ID)
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := clients.RecoveryServices.ReplicationPoliciesClient.Get(ctx, *id)
+	if err != nil {
+		return nil, fmt.Errorf("reading site recovery replication policy (%s): %+v", id.String(), err)
```
Comment (stephybun) [inline internal/services/recoveryservices/site_recovery_replication_policy_hyperv_resource_test.go:90]:
```suggestion
		return nil, fmt.Errorf("retrieving %s: %+v", *id, err)
```

File: internal/services/recoveryservices/site_recovery_replication_policy_hyperv_resource_test.go [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+	}
+
+	model := resp.Model
+	if model == nil {
+		return nil, fmt.Errorf("reading site recovery replication policy (%s): model is nil", id.String())
+	}
+
+	return utils.Bool(model.Id != nil), nil
```
Comment (stephybun) [inline internal/services/recoveryservices/site_recovery_replication_policy_hyperv_resource_test.go:98]:
```suggestion
	return utils.Bool(resp.Model != nil), nil
```

File: website/docs/r/site_recovery_replication_policy_hyperv.html.markdown [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+subcategory: "Recovery Services"
+layout: "azurerm"
+page_title: "Azure Resource Manager: azurerm_site_recovery_replication_policy_hyperv"
+description: |-
+    Manages an Azure Site Recovery replication policy for HyperV on Azure.
+---
+
+# azurerm_site_recovery_replication_policy_hyperv
+
+Manages a Azure Site Recovery replication policy fir HyperV within a recovery vault. Replication policies define the frequency at which recovery points are created and how long they are stored.
```
Comment (stephybun) [inline website/docs/r/site_recovery_replication_policy_hyperv.html.markdown:11]:
```suggestion
Manages a Azure Site Recovery replication policy for HyperV within a Recovery Vault. Replication policies define the frequency at which recovery points are created and how long they are stored.
```

File: internal/services/recoveryservices/site_recovery_replication_policy_hyperv_resource.go [hunk 14]
```diff
@@ -0,0 +1,244 @@
...
+func (r ReplicationPolicyHyperVResource) Attributes() map[string]*schema.Schema {
+	return map[string]*schema.Schema{}
+}
+
+func (r ReplicationPolicyHyperVResource) ModelObject() interface{} {
+	return &ReplicationPolicyHyperVModel{}
+}
+
+func (r ReplicationPolicyHyperVResource) ResourceType() string {
+	return "azurerm_site_recovery_replication_policy_hyperv"
```
Comment (stephybun) [inline internal/services/recoveryservices/site_recovery_replication_policy_hyperv_resource.go:77]: Can we rename this to 
```suggestion
	return "azurerm_site_recovery_hyperv_replication_policy"
```

File: website/docs/r/site_recovery_replication_policy_hyperv.html.markdown [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+
+* `name` - (Required) The name of the replication policy. Changing this forces a new resource to be created.
+
+* `resource_group_name` - (Required) Name of the resource group where the vault that should be updated is located. Changing this forces a new resource to be created.
+
+* `recovery_vault_name` - (Required) The name of the vault that should be updated. Changing this forces a new resource to be created.
+
+* `recovery_point_retention_in_hours` - (Required) The duration in hours for which the recovery points need to be stored.
+
+* `application_consistent_snapshot_frequency_in_hours` - (Required) Specifies the frequency(in hours) at which to create application consistent recovery points.
```
Comment (stephybun) [inline website/docs/r/site_recovery_replication_policy_hyperv.html.markdown:50]: The unit of time is already in the name of the property
```suggestion
* `application_consistent_snapshot_frequency_in_hours` - (Required) Specifies the frequency at which to create application consistent recovery points.
```

**Review by stephybun — CHANGES_REQUESTED** [review]
Thanks @ziyeqf, left some comments in-line, once those are fixed up we can have another look and get this merged!

File: internal/services/recoveryservices/site_recovery_replication_policy_hyperv_resource.go (hunk 10 above)
Comment (ziyeqf) [reply internal/services/recoveryservices/site_recovery_replication_policy_hyperv_resource.go]: have renamed this func to `expandHyperVToAzurePolicyDetail`, and H2A means HyperV To Azure, A2A means Azure To Azure

**Review by katbyte — CHANGES_REQUESTED** [review]
Test are failing..
```
------- Stdout: -------
=== RUN   TestAccSiteRecoveryReplicationPolicyHyperV_basic
=== PAUSE TestAccSiteRecoveryReplicationPolicyHyperV_basic
=== CONT  TestAccSiteRecoveryReplicationPolicyHyperV_basic
    testcase.go:110: Step 1/2 error: Error running pre-apply refresh: exit status 1
        
        Error: Missing required argument
        
          on terraform_plugin_test.tf line 31, in resource "azurerm_site_recovery_hyperv_replication_policy" "test":
          31: resource "azurerm_site_recovery_hyperv_replication_policy" "test" {
        
        The argument "recovery_vault_id" is required, but no definition was found.
        
        Error: Unsupported argument
        
          on terraform_plugin_test.tf line 32, in resource "azurerm_site_recovery_hyperv_replication_policy" "test":
          32:   resource_group_name                                = azurerm_resource_group.test.name
        
        An argument named "resource_group_name" is not expected here.
        
        Error: Unsupported argument
        
          on terraform_plugin_test.tf line 33, in resource "azurerm_site_recovery_hyperv_replication_policy" "test":
          33:   recovery_vault_name                                = azurerm_recovery_services_vault.test.name
        
        An argument named "recovery_vault_name" is not expected here.
--- FAIL: TestAccSiteRecoveryReplicationPolicyHyperV_basic (2.37s)
FAIL

```

**Comment by ziyeqf** [comment]
@katbyte sorry for that, forgot to update test case...

test after udpate:

```
terraform-provider-azurerm ❯ TF_ACC=1 go test -v ./internal/services/recoveryservices -run=TestAccSiteRecoveryReplicationPolicyHyperV_ -timeout=600m                 [ tengzh/recovery/hyperv/policy][$][🐹 v1.19.5][⏱ 1m1s][11:21:20]
=== RUN   TestAccSiteRecoveryReplicationPolicyHyperV_basic
=== PAUSE TestAccSiteRecoveryReplicationPolicyHyperV_basic
=== RUN   TestAccSiteRecoveryReplicationPolicyHyperV_noSnapshots
=== PAUSE TestAccSiteRecoveryReplicationPolicyHyperV_noSnapshots
=== CONT  TestAccSiteRecoveryReplicationPolicyHyperV_basic
=== CONT  TestAccSiteRecoveryReplicationPolicyHyperV_noSnapshots
--- PASS: TestAccSiteRecoveryReplicationPolicyHyperV_noSnapshots (474.58s)
--- PASS: TestAccSiteRecoveryReplicationPolicyHyperV_basic (475.15s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/recoveryservices      476.410s

```

**Review by stephybun — APPROVED** [review]
Thanks @ziyeqf LGTM 🍀

**Pull request merged**